[inih] Build with Meson + update to r57

DownerCase opened the pull request:
<!-- If your PR fixes issues, please note that here by adding "Fixes #NNNNNN." for each fixed issue on separate lines. -->
Follows on from #17388
Fixes #17385
Done in preparation to continue #31722 and properly address [this review comment](https://github.com/microsoft/vcpkg/pull/31722#discussion_r1246072471)

If this PR updates an existing port, please uncomment and fill out this checklist:

- [X] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md)
- [X] SHA512s are updated for each updated download
- [X] The "supports" clause reflects platforms that may be fixed by this new version
- [X] Any fixed [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt) entries are removed from that file.
- [X] Any patches that are no longer applied are deleted from the port's directory.
- [X] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [X] Only one version is added to each modified port's versions file.

Repo I've used to verify proper usage: https://github.com/DownerCase/vcpkg_verifier/tree/inih

Writing a CMake package for this by hand was an interesting experience. There must be a better way to do this because it won't work for anything more complex.

DownerCase committed:
[inih] Use meson build, bump version

DownerCase committed:
[inih] Versioning

DownerCase commented:
@microsoft-github-policy-service agree

DownerCase committed:
Specify c++11 to meson to fix osx?

DownerCase committed:
[inih] Version

DownerCase commented:
Yay! That actually worked 😄 

> Writing a CMake package for this by hand was an interesting experience. There must be a better way to do this because it won't work for anything more complex.

Well meson does have a [CMake module](https://mesonbuild.com/CMake-module.html#cmake-configuration-files) that can write a version file, but beyond that it's [not super useful](https://github.com/mesonbuild/meson/issues/7632)

JonLiu1993 commented:
@DownerCase, Thanks for your PR, please restore the newline type for portfile.cmake and CMakeLists.txt.

JonLiu1993 commented:
Note: I will be converting your PR to draft status. When you respond, please revert to "ready for review". That way, I can be aware that you've responded since you can't modify the tags.

DownerCase committed:
Restore line endings

JonLiu1993 commented:
Please run command `./vcpkg x-add-version inih --overwrite-version` and commit agagin.

DownerCase committed:
Version

dg0yt commented:
IMO the port should get rid of `vcpkg_cmake_...` and just directly configure and install the desired (maybe simplified ) cmake config file. (I would also drop the corresponding version file. This is vcpkg, there are manifests.)

DownerCase commented:
I mean sure, I considered that at one point but it felt wrong somehow to do all that in the portfile so I used the CMakeList instead.
If that’s actually preferable, and I can see how it would be, I can change it no problem.

The version file was thrown in just to “complete” the package and because it was easy. I’m not attached to it.

If you have particular suggestions on how to simplify the cmake.in file I’m all ears. The annoying bit for me was that the meson build generates two library files, as opposed to the current port build which bundles them into one, requiring two CMake imported targets and all the duplication thereof.

JonLiu1993 commented:
Tesed usage successfully by `inih:x64-windows`:
````
    find_package(unofficial-inih CONFIG REQUIRED)
    target_link_libraries(main PRIVATE unofficial::inih::libinih )
````

dg0yt commented on ports/inih/unofficial-inihConfig.cmake.in:
```diff
@@ -0,0 +1,94 @@
+####  libinih  ####
+
+# Add library target (note: inih always has a static build in vcpkg).
+add_library(unofficial::inih::libinih STATIC IMPORTED)
```
IMO the imported target should be `UNKNOWN` instead of `STATIC`.

dg0yt commented on ports/inih/unofficial-inihConfig.cmake.in:
```diff
@@ -0,0 +1,94 @@
+
+if(INIH_WITH_INI_READER)
+  # Add library target (note: inih always has a static build in vcpkg).
+  add_library(unofficial::inih::inireader STATIC IMPORTED)
```
As above: `UNKNOWN`.

dg0yt commented on ports/inih/unofficial-inihConfig.cmake.in:
```diff
@@ -0,0 +1,94 @@
+###################
+
+# Check header and library files are present.
+foreach(file ${_IMPORT_CHECK_FILES} )
```
IMO this doesn't make sense now: You could just use add `REQUIRED` to the `find_library` calls.
(Okay, the loop also checks for `INIReader.h`, but do we really need these late checks? The package was already built, installed and cached before this CMake config is ever loaded.)

dg0yt commented on ports/inih/usage:
```diff
@@ -0,0 +1,4 @@
+inih provides CMake targets:
+    find_package(unofficial-inih CONFIG REQUIRED)
+    target_link_libraries(main PRIVATE unofficial::inih::libinih @inireader_usage@)
```
Enumerating all libs/targets is what the usage heurisitics does because it doesn't know better. Users will follow it blindly, and then they always need the optional port feature. Cf. `ports/geos/usage` for how to explain different libs. (And maybe the optional feature should just be made a standard thing, for simplicity. It doesn't add any dependencies.)

DownerCase replied:
Yeah I was looking to recreate the heuristic message but with something appended to the end. I don't think there is a way to do that?

DownerCase replied:
Optional feature is now default

dg0yt commented on ports/inih/usage:
```diff
@@ -0,0 +1,4 @@
+inih provides CMake targets:
+    find_package(unofficial-inih CONFIG REQUIRED)
```
```suggestion
inih provides CMake targets:

    find_package(unofficial-inih CONFIG REQUIRED)

```

dg0yt commented on ports/inih/usage:
```diff
@@ -0,0 +1,4 @@
+inih provides CMake targets:
+    find_package(unofficial-inih CONFIG REQUIRED)
+    target_link_libraries(main PRIVATE unofficial::inih::libinih @inireader_usage@)
+Alternatively use via PkgConfig
```
This sentence stops unexpectedly (period, line break, details).

With official pkgconfig support, the unofficial cmake stuff is much less attractive for software which isn't built with vpckg only. Cf. `ports/poppler/usage` for how to show the details.

DownerCase replied:
Sorry for the off topic... but what is required to make that method work? I was always told it couldn't find the pkgconfig executable despite having pkgconf as a host dependency in the manifest and the executable in my tools folder.

DownerCase replied:
Unofficial CMake config has been removed.

DownerCase commented:
I'll address the review comments later today.

DownerCase committed:
Address review comments

- Removed CMake package versioning
  - Wasn't that useful as upstream don't use semver
- Use straight configure_file to install CMake package config
- Make cpp feature default
- Prefer PkgConfig usage

DownerCase committed:
Version

DownerCase committed:
Actually delete the CMakeList

DownerCase committed:
Version

JonLiu1993 commented on ports/inih/usage:
```diff
@@ -0,0 +1,12 @@
+The package inih can be imported via CMake FindPkgConfig module:
+
+    find_package(PkgConfig)
```
@DownerCase, Does the first method need to add a dependency on pkgconfig in vcpkg.json? If the usage is added, the test can pass, otherwise it will always show that:
````
could not find pkgconfig (missing: pkg_config_executable)
````

DownerCase replied:
Yeah so you need to have the pkg config executable available somehow for it to work. The usage text is identical to what is display for other ports with pkgconfig ability.
My understanding is that the host machine either needs the executable to be on system path or that you add "pkgconf" to your manifest and tell CMake where to look for it (this bit I do not know how to do).

When I was looking around on this pkgconf stuff other reviewers seemed pretty emphatic on *not* adding pkgconf as a dependency if it is not strictly required to build/use the package. I did not understand fully though.

JonLiu1993 replied:
I confirm that there is no need to add pkgconf dependencies, we install them ourselves when testing.

JonLiu1993 commented:
feature cpp tested successfully in the following triplet:
- x86-windows
- x64-windows
- x64-windows-static

Tested usage successfully by `inih:x64-windows`:
````
The package inih can be imported via CMake FindPkgConfig module:

    find_package(PkgConfig)
    pkg_check_modules(INIH REQUIRED IMPORTED_TARGET inih INIReader)

    target_link_libraries(main PRIVATE PkgConfig::INIH)

Alternatively, the package inih provides unofficial CMake targets:

    find_package(unofficial-inih CONFIG REQUIRED)

    target_link_libraries(main PRIVATE unofficial::inih::libinih unofficial::inih::inireader)
````

dg0yt reviewed (commented):
AFAIU the package offers two things: a C library, and a C++ class. Even if the latter uses the former, users will only access one or another *directly*.
IMO `usage` should list these patterns separately, e.g.

~~~
inih can be imported via CMake FindPkgConfig module:

    find_package(PkgConfig)
    # C API
    pkg_check_modules(inih REQUIRED IMPORTED_TARGET inih)
    target_link_libraries(main PRIVATE PkgConfig::inih)
    # C++ API
    pkg_check_modules(inih REQUIRED IMPORTED_TARGET INIReader)
    target_link_libraries(main PRIVATE PkgConfig::INIReader)
~~~~

Simlilar for the unofficial cmake config.

DownerCase commented:
@dg0yt Your understanding is correct. I'll make the change but this usage is becoming rather long. Is it ok for the C++ API usage to still be present even if the feature has been turned off and it is not present?

DownerCase committed:
separate C and C++ library usage

DownerCase committed:
Version

dg0yt commented:
> I'll make the change but this usage is becoming rather long. 

I'm aware of the length. That's why I'm against promoting unofficial usage when there is an official pattern. 
(But I do understand the limitations of pkg-config for multi-config builds.)

> Is it ok for the C++ API usage to still be present even if the feature has been turned off and it is not present?

I think I made the point before: Remove the optional feature and always install the C++ API. I see no benefit from making it optional in this package.

DownerCase commented:
> > Is it ok for the C++ API usage to still be present even if the feature has been turned off and it is not present?
> 
> I think I made the point before: Remove the optional feature and always install the C++ API. I see no benefit from making it optional in this package.

You did. I think I misunderstood and made it a default-option. IMO what we have now is good enough; inih is a C library first and they build INIReader as an optional feature, makes sense to me to reflect that. Something to revisit next update perhaps.

DownerCase commented:
Great ci flaked out, 503 when downloading vcpkg tool. Rerunning x64_windows_static_md is needed.

JonLiu1993 commented:
Tested usage successfully by inih:x64-windows:
````
The package inih can be imported via CMake FindPkgConfig module:

    find_package(PkgConfig)

    # C API
    pkg_check_modules(INIH REQUIRED IMPORTED_TARGET inih)
    target_link_libraries(main PRIVATE PkgConfig::INIH)

    # C++ API (Requires "cpp" feature)
    pkg_check_modules(INIReader REQUIRED IMPORTED_TARGET INIReader)
    target_link_libraries(main PRIVATE PkgConfig::INIReader)

Alternatively, the package inih provides unofficial CMake targets:

    find_package(unofficial-inih CONFIG REQUIRED)

    # C API
    target_link_libraries(main PRIVATE unofficial::inih::libinih)

    # C++ API (Requires "cpp" feature)
    target_link_libraries(main PRIVATE unofficial::inih::inireader)
````

JavierMatosD commented on ports/inih/usage:
```diff
@@ -0,0 +1,21 @@
+The package inih can be imported via CMake FindPkgConfig module:
+
+    find_package(PkgConfig)
+
+    # C API
+    pkg_check_modules(INIH REQUIRED IMPORTED_TARGET inih)
+    target_link_libraries(main PRIVATE PkgConfig::INIH)
+
+    # C++ API (Requires "cpp" feature)
+    pkg_check_modules(INIReader REQUIRED IMPORTED_TARGET INIReader)
+    target_link_libraries(main PRIVATE PkgConfig::INIReader)
+
+Alternatively, the package inih provides unofficial CMake targets:
+
+    find_package(unofficial-inih CONFIG REQUIRED)
+
+    # C API
+    target_link_libraries(main PRIVATE unofficial::inih::libinih)
+
+    # C++ API (Requires "cpp" feature)
+    target_link_libraries(main PRIVATE unofficial::inih::inireader)
```
I don't think we should be giving instructions for using pckgconfig through cmake. I think it should just say something like, "if you are using CMake, the package provides unofficial CMake targets ...". If you are using pckgconfig here's the name. 

```suggestion
The package inih provides unofficial CMake targets:

    find_package(unofficial-inih CONFIG REQUIRED)

    # C API
    target_link_libraries(main PRIVATE unofficial::inih::libinih)

    # C++ API (Requires "cpp" feature)
    target_link_libraries(main PRIVATE unofficial::inih::inireader)

Alternatively, if you are using pckgconfig use the name "inih" for the C API and "inireader" for the C++ API

```

DownerCase replied:
I’ve just been going with established patterns, sounds like you guys need to form a consensus and guidance on how to handle usage texts and apply that uniformly. Something not in the scope of this PR which was to update a small package.

DownerCase replied:
@dg0yt As the initial reviewer would you like to weigh in on this? You expressed a preference for the pkgconfig method previously as it is the officially supported method which I agree with and is why I put the unofficial CMake target second as the alternative usage.

Having full instructions for official usage makes sense to me and is what I would expect to see. 

From my perspective as a user there are two things that matter.
1. Does the package build cleanly for my case. CI suggests yes.
2. Can I read the usage text and get it working in my project. JonLiu has verified that it can.

dg0yt replied:
> You expressed a preference for the pkgconfig method previously as it is the officially supported method which I agree with and is why I put the unofficial CMake target second as the alternative usage.

Give the feedback from vcpkg's @JavierMatosD, I don't want to ask for more changes. But I do have a clear `<opinion>`: Don't add or announce unofficial CMake config (vcpkg lock-in) when there is official pkg-config.
I could accept the short form of announcing the pkgconfig option, but IMO it is the primary form (official, no lock-in), not the alternative.`</opinion>`. From that POV, I can't approve, but it doesn't matter anyways.

> Can I read the usage text and get it working in my project. JonLiu has verified that it can.

The usual quick validation of usage just scratches the surface. It detects some logical bugs, and this is a good thing. But normally it can't detect semantical defects in usage hints which are taken literally from the uninformed heuristics. So when adding `usage`, better spend an extra loop on reviewing it from the perspective of an uninitiated user.

DownerCase replied:
Ok fine. @JavierMatosD how about this. We drop the unofficial CMake support completely and only retain the first half of the current usage text? Pkgconfig is made available by upstream and users must use that to use this package.

Downstream would just have to deal with the unofficial target disappearing on them. I know this at least this will affect minio-cpp. If this goes forward I will sort out the change with them.

DownerCase replied:
Unofficial CMake support has been dropped in favor of pkgconfig. If people have trouble migrating, tag me in and I'll help.

DownerCase committed:
Remove unofficial CMake config

DownerCase commented:
Alright new state of play. 
- No unofficial CMake target because upstream provides a pkg config output. 
  - I nominate myself as responsible if downstream/users have issues. Only aware of minio-cpp currently.
- Usage:
  - Is consistent with other pkgconf only packages.
  - Lists both available packages and their use-case
  - Notes that INIReader requires the "cpp" feature which is enabled by default.
- Now supports dynamic linkage! 🥳
  - I tested the usage and although upstream has the soversion hard-coded as '0' (inih-0.dll) it did work.

DownerCase committed:
Revert "Remove unofficial CMake config"

This reverts commit 94ff6a903203912b96012c8c3479e6ccf9cc2308.

DownerCase committed:
Take Javier's suggestion

DownerCase commented:
My respect for all the vcpkg maintainers has grown hugely after getting a taste of the madness you have to deal with. Thank you for all the hard work to make this all usable. 
For some reason I still feel compelled to want to continue contributing though; hmm.

Anyway, hopefully this PR is almost done now...
